### ISSUE TITLE
Runtime bug: Simultaneous keypresses not recognized all the time

### ISSUE THREAD

**Issue by fatherom123** [dxx-rebirth/dxx-rebirth#514]
Windows 10
20200601-win
D1X-rebirth
single player

### Description

I use just the keyboard when I play. I map most of the controls to the numeric keypad. I use the up and down arrow for accel and reverse. I use 8, 4, 6 and 2 on the numeric keyboard for turning.

I just installed everything tonight (D1X-rebirth over top steam installation). Game looks great, plays fine, BUT:

When I'm accel/reverse, the turn controls only work some of the time. It's as if pressing multiple keys simultaneously isn't being recognized...as such, the gameplay is impossible for how I'm used to playing. Is there some config step I missed?

**Comment by AlumiuN** [comment]
I can't replicate this; do the controls work properly if you run Descent in Dosbox (i.e. the default Steam installation)?

**Comment by fatherom123** [comment]
Hi there,

Thanks for the quick response...I'm hoping this is something simple and I can relive my youth playing this game. :)

Some data points:

1) the default steam installation plays fine (no problem pressing multiple keys and having them be recognized)

2) the default steam installation won't completely remove the HUD (I pressed f3 multiple times, but I still have 1 of the 2 HUDs showing, not the total full screen minimal hud) - not sure if this matters at all...figured I'd mention it. dxx-rebirth removes the HUD just fine.

3) I basically placed (and overwrote one file) all the files from the d1x-rebirth zip over top of my steam installation and launched the game using the d1x-rebirth.exe.

Thanks again,

Chris

**Comment by vLKp** [comment]
I suggest starting the game with the `-debug` command line option, which should make it log keyup / keydown events.  Then, reproduce the problem, exit the game, and check the log for messages of the form `Sending event EVENT_KEY_`.  Look for whether the game sent key events for all the flight keys, or only the keys that you observed the ship actually handle.  Report back which occurred.  This will differentiate between a problem where the key was detected, but not respected, versus a problem where the key was not detected at all.

**Comment by fatherom123** [comment]
Thanks again.

I've got it boiled down to a pretty reproducible scenario. Here's the log chunk you asked for:

`11:42:59.179 Sending event EVENT_KEY_COMMAND:      UP
11:43:01.893 Sending event EVENT_KEY_RELEASE:      UP
11:43:02.426 Sending event EVENT_KEY_COMMAND:      ESC`

The up is me accelerating forward...at the same time I was pressing the 8 on the numeric keyboard and my ship wasn't turning. And there's no event for me pressing the 8 key. I then pressed ESC to exit the game.

**Comment by fatherom123** [comment]
OK, I think I solved it. Was doing something on my daughter's laptop and realized I couldn't play Descent on it since its keyboard doesn't have a separate numeric keypad. Then the light bulb came on.

Went back to my desktop PC and realized num lock was off. Once I turned it on, the controls in d1x-rebirth seemed fine.

It's interesting, since I'm pretty sure num lock was off when I played the native Steam version (which worked fine).

**Comment by vLKp** [comment]
Based on that debug output, in the failure case, the relevant event was never sent to the game, so it could not react to it.

For me, the keybinding PAD8 is recognized regardless of whether Numlock is enabled.  It would be interesting to see your results for this test:
- Open the keyboard controls configuration area (Main Menu -> Options -> Controls -> Customize Keyboard)
- In one of the boxes for Pitch Forward, press Enter to change the binding.  Press your numpad 8 while Numlock is enabled.  Note whether it changes, and what symbol it shows.  For me, I get `PAD8`.
- Repeat this change, but with Numlock disabled.  What happens?  Does the game remain in prompt mode, as if it never saw the key?  Does it record some other symbol?  For me, I get the same results as the Numlock enabled case.

Since this appears to be a configuration error outside the game's control, I am closing this issue so that future readers do not think there is a bug here.  You're welcome to use the issue to continue to discuss this.  If you find a reproducible bug, reopen the issue.

If you think the game could do better to help users, we can discuss that too.  I could see having a Numlock indicator somewhere, but I am reluctant to clutter the display with information that may be irrelevant to users.

**Comment by fatherom123** [comment]
Follow up:

Whether num lock is on or off, the Customize Keyboard screen always shows "Pad 8" for the inputted key. 

Even if configuring the keyboard with num lock off, once I start playing the game turning num lock back to on resume normal expected gameplay behavior.

**Comment by vLKp** [comment]
That is strange.  Could you run with `-debug` and check what the kconfig screen reports with Numlock disabled?  I would have expected that kconfig and the game should behave the same: either both require Numlock enabled or both ignore Numlock entirely.  Instead, your results show that it matters in one case and is ignored in the other.

**Comment by fatherom123** [comment]
OK, tried what you asked. Here's the log (num lock off initially, turned it on halfway through doing key config)

`10:14:17.683 Sending event EVENT_KEY_RELEASE:      ENTER
10:14:21.510 Sending event EVENT_KEY_COMMAND:      ENTER
10:14:21.587 Sending event EVENT_KEY_RELEASE:      ENTER
10:14:22.420 Sending event EVENT_KEY_COMMAND:      PAD8
10:14:22.505 Sending event EVENT_KEY_RELEASE:      PAD8
10:14:23.712 Sending event EVENT_KEY_COMMAND:      ENTER
10:14:23.788 Sending event EVENT_KEY_RELEASE:      ENTER
10:14:24.480 Sending event EVENT_KEY_COMMAND:      PAD8
10:14:24.568 Sending event EVENT_KEY_RELEASE:      PAD8
10:14:25.963 Sending event EVENT_KEY_COMMAND:      NMLCK
10:14:26.967 Sending event EVENT_KEY_COMMAND:      ENTER
10:14:27.054 Sending event EVENT_KEY_RELEASE:      ENTER
10:14:27.649 Sending event EVENT_KEY_COMMAND:      PAD8
10:14:27.735 Sending event EVENT_KEY_RELEASE:      PAD8
10:14:28.637 Sending event EVENT_KEY_COMMAND:      ENTER
10:14:28.704 Sending event EVENT_KEY_RELEASE:      ENTER
10:14:29.231 Sending event EVENT_KEY_COMMAND:      PAD8
10:14:29.317 Sending event EVENT_KEY_RELEASE:      PAD8
10:14:33.163 Sending event EVENT_KEY_COMMAND:      ESC`

Basically, it's always registering pad8.

So here's something that may have not been clear when I initially described the issue. When numlock is off, it's not that pad8 is ignored...it works about 50% of the time. It's like I have to repeatedly press the key for it to be recognized, especially if I'm also accel/reverse at the same time.

**Comment by fatherom123** [comment]
Another data point: was just playing a bit more...and noticed: even with num lock on, I can't seem to press 3 keys simultaneously (i was accelerating and firing lasers and tried to turn, and the turn (pad 8) didn't work.

I wonder if this is related to USB keyboard rollover and max key presses allowed

**Comment by fatherom123** [comment]
I went to:

http://gadzikowski.com/nkeyrollover.html

and my keyboard won't register UPARROW, PAD8 and PAD0 at the same time. (PAD0 is what I use for firing and it wouldn't be recognized if the other two keys were already pressed/engaged).

**Comment by vLKp** [comment]
Based on the last two comments, would it be correct to say that this appears to be a problem with your keyboard in general, and is not specific to Rebirth?  I was generally aware that keyboards had a maximum number of concurrent keys, but I don't think I've ever seen one that had such a low limit before.

**Comment by fatherom123** [comment]
Yeah, I think it's not a rebirth-specific issue.
